File: src/main/java/net/dblsaiko/illuminate/test/BlockLight.java
package net.dblsaiko.illuminate.test;

import com.mojang.blaze3d.systems.RenderSystem;
import net.dblsaiko.illuminate.client.IlluminateClient;
import net.dblsaiko.illuminate.client.api.Light;
import net.minecraft.block.BlockState;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.texture.TextureManager;
import net.minecraft.state.property.Properties;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import org.joml.Vector3f;
import org.joml.Vector3fc;

public class BlockLight implements Light {
    private static final float Z_NEAR = (float) (Math.sqrt(2f) / 2);

    private final World world;
    private final BlockPos pos;

    private final Vector3f posBuf = new Vector3f();
    private boolean wait = true;
    private float yaw;

    public BlockLight(World world, BlockPos pos) {
        this.world = world;
        this.pos = pos;
    }

    @Override
    public int tex() {
        return getTex();
    }

    @Override
    public Vector3fc pos() {
        this.posBuf.set(this.pos.getX() + 0.5f, this.pos.getY() + 0.5f, this.pos.getZ() + 0.5f);
        return this.posBuf;
    }

    @Override
    public float yaw() {
        return this.yaw;
    }

    @Override
    public float fov() {
        return 35f;
    }

    @Override
    public float aspect() {
        return 768 / 576f;
    }

    @Override
    public float near() {
        return Z_NEAR;
    }

    @Override
    public void prepare(float delta) {
        BlockState state = this.world.getBlockState(this.pos);

        if (state.getBlock() != IlluminateTest.instance().blocks.projector) {
            if (!this.wait) {
                IlluminateClient.instance().removeLight(this);
            }

            return;
        } else {
            this.wait = false;
        }

        this.yaw = state.get(Properties.HORIZONTAL_FACING).asRotation();
    }

    public static int getTex() {
        TextureManager tm = MinecraftClient.getInstance().getTextureManager();
        Identifier id = IlluminateTest.id("textures/test.png");
        tm.bindTexture(id);
        RenderSystem.bindTexture(0);

        return tm.getTexture(id).getGlId();
    }
}

File: src/main/kotlin/client/test/BlockLight.kt
package therealfarfetchd.illuminate.client.test

import com.mojang.blaze3d.systems.RenderSystem
import net.dblsaiko.illuminate.client.IlluminateClient
import net.dblsaiko.illuminate.client.api.Light
import net.minecraft.client.MinecraftClient
import net.minecraft.state.property.Properties
import net.minecraft.util.Identifier
import net.minecraft.util.math.BlockPos
import org.joml.Vector3f
import therealfarfetchd.illuminate.ModID
import therealfarfetchd.illuminate.client.glwrap.WGlTexture2D
import therealfarfetchd.illuminate.common.init.Blocks
import kotlin.math.sqrt

class BlockLight(val bp: BlockPos) : Light {

  override fun tex() = getTex().id

  override fun pos(): Vector3f = Vector3f(bp.x + 0.5f, bp.y + 0.5f, bp.z + 0.5f)

  private var yaw = 80f
  override fun yaw(): Float = yaw

  override fun fov(): Float = 35f

  override fun aspect(): Float = 768/576f

  override fun near(): Float = sqrt(2f) / 2

  private var wait = true

  override fun prepare(delta: Float) {
    val state = MinecraftClient.getInstance().world!!.getBlockState(bp)

    if (state.block != Blocks.Projector) {
      if (!wait) IlluminateClient.instance().removeLight(this)
      return
    } else {
      wait = false
    }

    yaw = -state.get(Properties.HORIZONTAL_FACING).asRotation()
  }

  companion object {
    fun getTex(): WGlTexture2D {
      val tm = MinecraftClient.getInstance().textureManager
      val identifier = Identifier(ModID, "textures/test.png")
      tm.bindTexture(identifier)
      RenderSystem.bindTexture(0)

      return WGlTexture2D(tm.getTexture(identifier)!!.glId)
    }
  }

}
